src/components/HostsForm/PuppetConfig/Custom/PuppetEnvSelectInput.js

```diff
@@ -1,26 +1,35 @@
 import React, { useContext } from 'react'
 import T from 'i18n-react'
-import { useQuery } from '@apollo/react-hooks'
-import { get } from 'lodash'
+import { find, get } from 'lodash'
 import { HostsFormContext } from 'lib/Context'
 import SelectInput from 'components/HostsForm/SelectInput'
-import PUPPET_ENVS_QUERY from 'graphql/queries/puppetEnvs';
 
-const PuppetEnvSelectInput = ({...attrs}) => {
+const PuppetEnvSelectInput = ({
+  puppetEnvironments: { loading, data },
+  ...attrs
+}) => {
   const {
     updateAttribute,
     attributes: {
+      puppetclassIds = [],
       puppetEnvId
     }
   } = useContext(HostsFormContext)
 
-  const { loading, data } = useQuery(PUPPET_ENVS_QUERY, {
-    fetchPolicy: 'cache-and-network'
-  })
-  const puppetEnvs = get(data, 'environments.edges', []).map(({ node: { id, name }}) => ({ id, name }))
+  const puppetEnvs = get(data, 'environments.edges', [])
+    .map(({
+      node: {
+        id,
+        name,
+        puppetclasses: { edges: puppetclasses }
+      }
+    }) => ({ id, name, puppetclasses: puppetclasses.map(({ node: { id }}) => id) }))
+
+  const currentPuppetEnv = find(puppetEnvs, ['id', puppetEnvId]) || {}
+  const newPuppetclassIds = puppetclassIds.filter((id) => get(currentPuppetEnv, 'puppetclasses', []).includes(id))
```

Comment:
we need to filter out puppet classes that do not belong to the selected puppet environment